samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemAdmin.scala

```diff
@@ -554,14 +575,34 @@ class KafkaSystemAdmin(
     topicMetadata.partitionsMetadata.isEmpty
   }
 
+  /**
+    * @inheritdoc
+    *
+    * Delete records up to (and including) the provided ssp offsets for all system stream partitions specified in the map
+    * This only works with Kafka cluster 0.11 or later. Otherwise it's a no-op.
+    */
+  override def deleteMessages(offsets: util.Map[SystemStreamPartition, String]) {
+    deleteMessagesCalled = true
+
+    if (!running) {
+      throw new SamzaException(s"KafkaSystemAdmin has not started yet for system $systemName")
+    }
+    if (deleteMessagesEnabled) {
+      val nextOffsets = offsets.asScala.toSeq.map { case (systemStreamPartition, offset) =>
+        (new TopicPartition(systemStreamPartition.getStream, systemStreamPartition.getPartition.getPartitionId), offset.toLong + 1)
+      }.toMap
+      adminClient.deleteRecordsBefore(nextOffsets)
```

Comment:
Just out-of-curiosity, what if the adminClient failed during the SystemAdmin's life-cycle? Do we throw-away the failed adminClient and recreate a new one? If not, how do we recover? Or are we saying that the exception from deleteMessages() would definitely fail the SamzaContainer? I would prefer that this deleteMessages() can be a best-effort operation. i.e. if there is a network connection issue that does not allow deleteMessages() now, in next round of commit, when the network connection issue is gone, we should be able to deleteMessages() right?

Reply:
Good point. AdminClient has a background thread that keeps trying to send record and reconnect to broker if needed etc. If AdminClient is disconnected from broker, adminClient.deleteRecordsBefore() will still enqueue request into a request queue and return a future. If AdminClient connects to broker before timeout, then request will be sent and executed. Otherwise, a TimeoutException will be returned if user calls future.get(). Samza would not notice any failure since we do not call get() on the future returned by adminClient.deleteRecordsBefore().